CRM/Contribute/Tokens.php

```diff
@@ -55,21 +56,33 @@ public function getCurrencyFieldName() {
    */
   protected function getRelatedTokens(): array {
     $tokens = [];
-    if (!in_array('ContributionRecur', array_keys(\Civi::service('action_object_provider')->getEntities()))) {
-      return $tokens;
-    }
-    $hiddenTokens = ['modified_date', 'create_date', 'trxn_id', 'invoice_id', 'is_test', 'payment_token_id', 'payment_processor_id', 'payment_instrument_id', 'cycle_day', 'installments', 'processor_id', 'next_sched_contribution_date', 'failure_count', 'failure_retry_date', 'auto_renew', 'is_email_receipt', 'contribution_status_id'];
-    $contributionRecurFields = ContributionRecur::getFields(FALSE)->setLoadOptions(TRUE)->execute();
-    foreach ($contributionRecurFields as $contributionRecurField) {
-      $tokens['contribution_recur_id.' . $contributionRecurField['name']] = [
-        'title' => $contributionRecurField['title'],
-        'name' => 'contribution_recur_id.' . $contributionRecurField['name'],
+
+    if (in_array('ContributionPage', array_keys(\Civi::service('action_object_provider')->getEntities()))) {
```

Comment:
This is getting a little repetitive, would it make more sense to define an array like [ContributionPage => [...], ContributionRecur => [...]] and loop over it here?

Reply:
@larssandergreen it's border line - probably OK for now

FYI one of the reason we added the (not really used as yet) `usage` key to the DAO was the idea we would start to declare which fields to render as tokens within it. Here you have just hard-coded a single field - which is probably OK for now
```
    <usage>
      <import>false</import>
      <export>true</export>
      <duplicate_matching>false</duplicate_matching>
    </usage>
```

Reply:
Also `if (in_array('ContributionPage', array_keys(\Civi::service('action_object_provider')->getEntities()` is always true if if `(in_array('ContributionRecur', array_keys(\Civi::service('action_object_provider')->getEntities()` or indeed if Contribute is enabled- @seamuslee001 added this as part of https://github.com/civicrm/civicrm-core/commit/e714cfff6ee934158cd06c90da723762ef36c67c - but I suspect it was required for the for Member_Tokens & not this one

Both require civicontribute extension to be enabled - but

Reply:
Ah - we need to move the class to the extension I guess - that way we don't need to check the extension is enabled - how scary is that to do? @seamuslee001

Reply:
I guess for now we can just keep this  - cos it will cover the ContributionPage too.. 
````
  if (!in_array('ContributionRecur', array_keys(\Civi::service('action_object_provider')->getEntities()))) {
      return $tokens;
    }
```

Reply:
So do we need this check at all? I can see that it makes sense for Member tokens, but for Contribution tokens, wouldn't this only be called if CiviContribute is enabled?

Reply:
@larssandergreen I think it would only be called when CiviContribute is enabled - but it might be still registered if not - so it might be the intervention is required when registering the token listener & then I guess there could be some issues when the extension is first disabled

Reply:
OK, so maybe we just check `if (in_array('Contribution', array_keys(\Civi::service('action_object_provider')->getEntities()`, to keep it simple and make it more clear what the check is doing?

Reply:
@larssandergreen sounds good & maybe a comment to cover the above discussion

Reply:
Done